LSP crashing with `Compiler is already in-use` error

```
14:32:11.902	INFO	CodeActionProvider codeActionsForCursor	Find code actions at /workspaces/bazel/bazel/src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java(884)...
14:32:11.904	SEVERE	LSP connect	Compiler is already in-use!
java.lang.RuntimeException: Compiler is already in-use!
	at org.javacs.ReusableCompiler.getTask(ReusableCompiler.java:111)
	at org.javacs.CompileBatch.batchTask(CompileBatch.java:105)
	at org.javacs.CompileBatch.<init>(CompileBatch.java:30)
	at org.javacs.JavaCompilerService.doCompile(JavaCompilerService.java:76)
	at org.javacs.JavaCompilerService.loadCompile(JavaCompilerService.java:67)
	at org.javacs.JavaCompilerService.compileBatch(JavaCompilerService.java:93)
	at org.javacs.JavaCompilerService.compile(JavaCompilerService.java:349)
	at org.javacs.JavaCompilerService.compile(JavaCompilerService.java:344)
	at org.javacs.action.CodeActionProvider.codeActionsForCursor(CodeActionProvider.java:37)
	at org.javacs.JavaLanguageServer.codeAction(JavaLanguageServer.java:515)
	at org.javacs.lsp.LSP.connect(LSP.java:390)
	at org.javacs.Main.main(Main.java:30)

[Error - 2:32:12 PM] Request textDocument/hover failed.
Error: Connection got disposed.
	at Object.dispose (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/main.js:904:25)
	at Object.dispose (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-languageclient/lib/client.js:74:35)
	at LanguageClient.handleConnectionClosed (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-languageclient/lib/client.js:2309:42)
	at LanguageClient.handleConnectionClosed (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-languageclient/lib/main.js:155:15)
	at closeHandler (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-languageclient/lib/client.js:2296:18)
	at CallbackList.invoke (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at closeHandler (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/main.js:240:26)
	at CallbackList.invoke (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/events.js:121:36)
	at StreamMessageReader.fireClose (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at Socket.<anonymous> (/home/vscode/.vscode-server/extensions/siliconsoldier.java-with-bazel-0.2.47/node_modules/vscode-jsonrpc/lib/messageReader.js:151:46)
	at Socket.emit (node:events:531:35)
	at Pipe.<anonymous> (node:net:338:12)
```

This manifested after a prior error.

```
14:32:08.520	INFO	LSP$1MessageReader peek	Cannot cancel request 2 because it has already started
14:32:11.871	SEVERE	LSP connect	java.lang.AssertionError
java.lang.IllegalStateException: java.lang.AssertionError
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.analyze(JavacTaskImpl.java:383)
	at org.javacs.CompileBatch.<init>(CompileBatch.java:43)
	at org.javacs.JavaCompilerService.doCompile(JavaCompilerService.java:76)
	at org.javacs.JavaCompilerService.loadCompile(JavaCompilerService.java:67)
	at org.javacs.JavaCompilerService.compileBatch(JavaCompilerService.java:93)
	at org.javacs.JavaCompilerService.compile(JavaCompilerService.java:349)
	at org.javacs.JavaCompilerService.compile(JavaCompilerService.java:344)
	at org.javacs.action.CodeActionProvider.codeActionsForCursor(CodeActionProvider.java:37)
	at org.javacs.JavaLanguageServer.codeAction(JavaLanguageServer.java:515)
	at org.javacs.lsp.LSP.connect(LSP.java:390)
	at org.javacs.Main.main(Main.java:30)
Caused by: java.lang.AssertionError
	at jdk.compiler/com.sun.tools.javac.util.Assert.error(Assert.java:155)
	at jdk.compiler/com.sun.tools.javac.util.Assert.checkNonNull(Assert.java:62)
	at jdk.compiler/com.sun.tools.javac.util.ListBuffer.append(ListBuffer.java:127)
	at jdk.compiler/com.sun.tools.javac.comp.Attr$1.visitYield(Attr.java:1637)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCYield.accept(JCTree.java:1677)
	at jdk.compiler/com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:50)
	at jdk.compiler/com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:58)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.lambda$visitSwitchExpression$7(Attr.java:1644)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.handleSwitch(Attr.java:1850)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitSwitchExpression(Attr.java:1629)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCSwitchExpression.accept(JCTree.java:1399)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitReturn(Attr.java:2475)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1728)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStat(Attr.java:736)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStats(Attr.java:755)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1439)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1092)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStat(Attr.java:736)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitIf(Attr.java:2249)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIf.accept(JCTree.java:1572)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStat(Attr.java:736)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStats(Attr.java:755)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitBlock(Attr.java:1439)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1092)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStat(Attr.java:736)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.visitMethodDef(Attr.java:1229)
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:916)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribTree(Attr.java:662)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribStat(Attr.java:736)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribClassBody(Attr.java:5655)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribClass(Attr.java:5543)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attribClass(Attr.java:5367)
	at jdk.compiler/com.sun.tools.javac.comp.Attr.attrib(Attr.java:5306)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.attribute(JavaCompiler.java:1359)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.attribute(JavaCompiler.java:1332)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.analyze(JavacTaskImpl.java:404)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$analyze$1(JavacTaskImpl.java:379)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.analyze(JavacTaskImpl.java:379)
	... 10 more
```

--- a/server/src/main/java/org/javacs/CompileBatch.java
+++ b/server/src/main/java/org/javacs/CompileBatch.java
@@ -96,7 +96,6 @@ class CompileBatch implements AutoCloseable {
 
     @Override
     public void close() {
-        borrow.close();
         closed = true;
     }
 

--- a/server/src/main/java/org/javacs/CompileBatch.java
+++ b/server/src/main/java/org/javacs/CompileBatch.java
@@ -42,6 +42,7 @@ class CompileBatch implements AutoCloseable {
             // You can get at `Element` values using `Trees`
             borrow.task.analyze();
         } catch (IOException e) {
+            borrow.close();
             throw new RuntimeException(e);
         }
     }
@@ -95,6 +96,7 @@ class CompileBatch implements AutoCloseable {
 
     @Override
     public void close() {
+        borrow.close();
         closed = true;
     }
 

--- a/server/src/main/java/org/javacs/JavaCompilerService.java
+++ b/server/src/main/java/org/javacs/JavaCompilerService.java
@@ -76,13 +76,13 @@ class JavaCompilerService implements CompilerProvider {
     private CompileBatch doCompile(Collection<? extends JavaFileObject> sources) {
         if (sources.isEmpty()) throw new RuntimeException("empty sources");
 
-        Set<Path> addFiles;
-        try (var firstAttempt = new CompileBatch(this, sources)) {
-            addFiles = firstAttempt.needsAdditionalSources();
-            if (addFiles.isEmpty()) return firstAttempt;
-            // If the compiler needs additional source files that contain package-private files
-            LOG.info("...need to recompile with " + addFiles);
-        }
+        var firstAttempt = new CompileBatch(this, sources);
+        var addFiles = firstAttempt.needsAdditionalSources();
+        if (addFiles.isEmpty()) return firstAttempt;
+        // If the compiler needs additional source files that contain package-private files
+        LOG.info("...need to recompile with " + addFiles);
+        firstAttempt.close();
+        firstAttempt.borrow.close();
 
         var moreSources = new ArrayList<JavaFileObject>();
         moreSources.addAll(sources);

--- a/server/src/main/java/org/javacs/JavaCompilerService.java
+++ b/server/src/main/java/org/javacs/JavaCompilerService.java
@@ -75,13 +75,15 @@ class JavaCompilerService implements CompilerProvider {
 
     private CompileBatch doCompile(Collection<? extends JavaFileObject> sources) {
         if (sources.isEmpty()) throw new RuntimeException("empty sources");
-        var firstAttempt = new CompileBatch(this, sources);
-        var addFiles = firstAttempt.needsAdditionalSources();
-        if (addFiles.isEmpty()) return firstAttempt;
-        // If the compiler needs additional source files that contain package-private files
-        LOG.info("...need to recompile with " + addFiles);
-        firstAttempt.close();
-        firstAttempt.borrow.close();
+
+        Set<Path> addFiles;
+        try (var firstAttempt = new CompileBatch(this, sources)) {
+            addFiles = firstAttempt.needsAdditionalSources();
+            if (addFiles.isEmpty()) return firstAttempt;
+            // If the compiler needs additional source files that contain package-private files
+            LOG.info("...need to recompile with " + addFiles);
+        }
+
         var moreSources = new ArrayList<JavaFileObject>();
         moreSources.addAll(sources);
         for (var add : addFiles) {

